### PR TITLE
#15 16/chore/implement dto and dummy endpoint

### LINKED ISSUE #15: [Backend] Implement data model (DTO)
## Why
About Web API, user might send request in unexpected format. So we need to check request format and type in run time. We can use Pydantic for this purpose.

## Definition of Done
Data Model a.k.a. DTO (Data Transfer Object) is defined.

## How
- Consider IO model and share to team members to reach agreement.
- Implement IO model using Pydantic `BaseModel` class.

### LINKED ISSUE #16: [Backend] Implement dummy endpoint
## Why
To test connection between frontend and backend, creating dummy endpoint (return fixed result) is the fastest way.

## Definition of Done
- Dummy endpoint for summarization which return fixed result is implemented. IO data should be validated by DTO which is created in #15 .

## How
- Implement dummy endpoint at `router.py`.
- Write test for view layer.

## Note
This ticket is blocked by #15 .

### PR DESCRIPTION
**Pull request by YoshikiKubotani**
## Issue URL

#15 #16 

## Change overview

- Add new endpoints for summarization with dummy controllers and use cases
  - As described [here](https://fastapi.tiangolo.com/ja/tutorial/request-files/#what-is-form-data), it cannot be done to send files with json requests at the same time due to the HTTP protocol limitation
  - I implemented two workarounds for that based on [Stackoverflow (Method2)](https://stackoverflow.com/questions/65504438/how-to-add-both-file-and-json-body-in-a-fastapi-post-request) - (a) and on [GitHub Issues](https://github.com/tiangolo/fastapi/issues/2387#issuecomment-906761427) - (b)
  - In (a), non-file data are given by query parameters. But specifying query parameters in the POST process does not seem to be a standard solution, so it might be best to avoid it
    - [httpx](https://www.python-httpx.org/quickstart/) does not mention specifying the params argument in POST (actually, it generates errors)
    - The [FastAPI documentation](https://fastapi.tiangolo.com/ja/tutorial/testing/#_5) also says to write query parameters directly
    - Also, this method has a problem with the OpenAI API key appearing in the URL
  - In (b), non-file data is given from the client as form data filled out in JSON type
- Add necessary Data Transfer Object(DTO)s as pydantic models
- Update mypy version from 1.2.0 to 1.3.0
- Implement unit tests for endpoints
  - Unit tests are also done both on (a) and (b) mentioned above

### Endpoint Unit Test

Run the following command.

```bash
# Run the following under Crux/environment/cpu
docker compose exec backend make test
```

### Architecture Check

Check the features separation is reasonable and follows [S.O.L.I.D Principles](https://medium.com/backticks-tildes/the-s-o-l-i-d-principles-in-pictures-b34ce2f1e898)

### Type Check

Run the following command.

```bash
# Run the following under Crux/environment/cpu
docker compose exec backend make mypy
```

## Note for reviewers

- If you explicitly wrap the endpoint returns in the `Response` class, the dynamic type checking by the pydantic model specified in response_model will not work. So I fixed the `Response` class in `default_response_class` and returned data as dictionary type.
- I tried to make the basis of the backend software architecture, but the folder structure remained unchanged under `backend/src`.  Please work on that task in #19 

## Questions

1. To solve the mypy error, I used pydantic models in parts other than `routers.py`, but is that correct?
2. The test is only passed when a fixed string (e.g., "This is a normal summary.") is returned from `controller.py`. When the logic is static, like Owan, the test can be configured to pass even after the logic is implemented. However, how should we configure the test for our case, where the solution changes dynamically (as the summary has a large variance)?
3. HTTP is supposed to be a stateless protocol, unlike FTP, etc. Cannot we send `multipart/form-data` and `application/json` consecutively as separate POST processes instead of simultaneously?

### CODE REVIEW

**Comment by gatheluck** [comment]
@YoshikiKubotani It looks like CI still raise error due to code format? Can you fix this?

**Comment by YoshikiKubotani** [comment]
@gatheluck 
That's because I added the basic template of the main logic under `backend/src/core_logic.py`, while it's not called.
and flake8 raises errors when functions and variables are defined but not used.

I think it's okay to ignore that since it would help us when we try to fully implement the logic in the future.

I've checked the type check and code format test were successfully passed.

File: applications/backend/src/dto.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+class UserIn(BaseModel):
+    openai_key: SecretStr
+    summary_format: Literal["normal", "three-point", "ochiai", "cvpaper"]
+
+
+class UserOut(BaseModel):
+    title: str
+    author: str
+    key_figure: str
+    summary_text: str
```
Comment (gatheluck) [inline applications/backend/src/dto.py]:
```suggestion
    summary_text: FormatNormal | FormatThreePoint | FormatOchiai | FormatCVPaper
```
Well... To me, type of `summary_text` should be like above...?

**Review by gatheluck — APPROVED** [review]
Let's merge this too if CI runs successfully.

**Pull request merged**